lib/LanguageServerCodeTransform/Model/NameImport/CandidateFinder.php

```diff
@@ -66,12 +66,16 @@ public function unresolved(TextDocumentItem $item): NameWithByteOffsets
     public function importCandidates(
         TextDocumentItem $item
     ): Generator {
-        $actions = [];
+        $added = [];
         foreach ($this->unresolved($item) as $unresolvedName) {
+            assert($unresolvedName instanceof NameWithByteOffset);
+            $nameString = (string)$unresolvedName->name();
+            if (array_search($nameString, $added, true) !== false) {
+                continue;
+            }
```

Comment:
You can use a dict instead:

```
$seen = [];
if (isset($seen[$hash])) {
    continue;
}

$seen[$hash] = true;
```

If the byte offest is not required then we can just return an array of strings....

Reply:
My reasoning is that byte offset is important for diagnostics only. That's why for one code action there can be multiple diagnostics (not sure we're doing that btw).

Reply:
Found one more optimisation, BTW. Wait for it.

Reply:
Done:
1. When providing diagnostics, there were candidate searches for each diagnostic. So cache the fact that there were or were not candidates for each name. Reuse it for next diagnostic for the same name and another offset.
2. When providing code actions, there were needless searches for candidates to display the diagnostic as a hint although we already _know_ there _are_ candidates. So use the same mechanism to not search but assume there are candidates and  the diagnostics were hints.

Reply:
Currently we're just not assigning all diagnostics that relate to a code action, which I view as a minor problem.